[BPK-1183] Flowtype popovers

matthewdavidson opened the pull request.

k0nserv commented on changelog.md:
```diff
@@ -6,6 +6,14 @@
+**Fixed:**
+- bpk-component-popover:
+- bpk-react-utils:
+  - `target` prop type is now correctly `oneOf([function, node])` instead of `oneOf([function, element])`
```
Flow finding bugs for us?

matthewdavidson replied:
Yeah!! I tried to use Element and flow was having none of it which made me realise I had use the wrong prop-type - strange that the prop-type didn't blow up in the first place 🤔 .

k0nserv replied:
👍

k0nserv commented on changelog.md:
```diff
@@ -6,6 +6,14 @@
+  - `target` prop type is now correctly `oneOf([function, node])` instead of `oneOf([function, element])`
+
+- bpk-component-tooltip:
+  - `target` prop type is now correct;y `node` instead of `element`
```
Ditto?

shaundon replied:
Small typo on `correctly` too!

matthewdavidson replied:
Yup!

k0nserv commented on packages/bpk-component-popover/index.js:
```diff
@@ -16,8 +16,13 @@
 import themeAttributes from './src/themeAttributes';
 
-export default BpkPopoverPortal;
+export type Props = PopoverProps;
```
So this is interesting, should we have a policy of doing this to prepare for a future where other people might use Flow?

matthewdavidson replied:
Yeah definitely - if we treat `stories.js` as a consumer then it was a necessity i.e. i couldn't have flow typed that file otherwise.

k0nserv replied:
Can we record a decision for that?

matthewdavidson replied:
Yeah sure!!

k0nserv commented on packages/bpk-component-popover/src/BpkPopover.js:
```diff
@@ -40,7 +42,19 @@ const bindEventSource = (source, callback) => event => {
-const BpkPopover = props => {
+export type Props = {
+  id: string,
+  onClose: Function,
```
Don't use `Function`. From the [docs](https://flow.org/en/docs/types/functions/)

> However, if you need to opt-out of the type checker, and don’t want to go all the way to any, you can instead use Function. Function is unsafe and should be avoided.

From the code it looks like the type should be 

```javascript
onClose: (source: $Values<EVENT_SOURCE>) => void,
```

matthewdavidson replied:
I ended up with:
```
onClose: (
    event: SyntheticEvent<>,
    props: {
      source: $Values<typeof EVENT_SOURCES>,
    },
  ) => void,
```

k0nserv replied:
👍

shaundon replied:
I like this; it's more readable :)

k0nserv commented on packages/bpk-component-popover/src/BpkPopover.js:
```diff
@@ -40,7 +42,19 @@ const bindEventSource = (source, callback) => event => {
+  label: string,
+  closeButtonText: string,
+  children: Node,
+  className: ?string,
```
Can we put maybe types last? Makes it easier to read

matthewdavidson replied:
Sure

k0nserv commented on packages/bpk-component-popover/src/BpkPopoverPortal-test.js:
```diff
@@ -99,7 +100,7 @@ describe('BpkPopoverPortal', () => {
     it('should render correctly with portalStyle added to portal component', () => {
       const shallowRenderer = Shallow.createRenderer();
-      const customStyle = { color: 'red' };
+      const customStyle: ?Object = { color: 'red' };
```
Why do we need to declare a type here? Per the [docs](https://flow.org/en/docs/types/objects/#toc-sealed-objects) flow should infer it

shaundon replied:
I got an error in native `BpkIcon` and had to [explicitly declare an object](https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-icon/src/BpkIcon.js#L90), maybe this is the same error.

k0nserv replied:
Hmm okay but I think that's because you have an unsealed object there. This object is sealed. Btw you can improve the type in BpkIcon with `const icons: { [string]: string } = {};`

matthewdavidson replied:
Yup flow moaned until I added that type so ...

k0nserv replied:
Strange, any reason it's not `const customStyle: { color: string } = { color: 'red' };`

matthewdavidson replied:
Pure laziness ... i didn't think it was worth the bother given that it's a test file.

shaundon replied:
@k0nserv you're right - my issue was due to an unsealed object. And thanks for the tip, I'll change that!

k0nserv commented on packages/bpk-component-popover/src/BpkPopoverPortal.js:
```diff
@@ -42,21 +44,64 @@ const onOpen = (popoverElement, targetElement) => {
+  target: Function | Node,
+  isOpen: boolean,
+  onClose: Function,
+  tetherOptions: Object,
```
I feel like we should list the exact possible options here, but since we might replace tether, let's defer that

matthewdavidson replied:
That was my thinking :P

k0nserv commented on packages/bpk-component-popover/src/BpkPopoverPortal.js:
```diff
@@ -42,21 +44,64 @@ const onOpen = (popoverElement, targetElement) => {
+  isOpen: boolean,
+  onClose: Function,
+  tetherOptions: Object,
+  portalStyle: ?HTMLElement,
```
This conflicts with `?Object` above and it isn't really a `HTMLElement` is it?

matthewdavidson replied:
Copypasta! good catch.

k0nserv commented on packages/bpk-component-popover/src/BpkPopoverPortal.js:
```diff
@@ -42,21 +44,64 @@ const onOpen = (popoverElement, targetElement) => {
+  tetherOptions: Object,
+  portalStyle: ?HTMLElement,
+  portalClassName: ?string,
+  renderTarget: ?Function,
```
Make a more specific function type

matthewdavidson replied:
ended up with `renderTarget: ?() => HTMLElement,`

k0nserv commented on packages/bpk-component-popover/src/BpkPopoverPortal.js:
```diff
@@ -42,21 +44,64 @@ const onOpen = (popoverElement, targetElement) => {
+  };
 
-  beforeClose(done) {
+  beforeClose = (done: Function) => {
```
We can have a better type here

matthewdavidson replied:
Ended up with `beforeClose = (done: () => void) => {`

k0nserv commented on packages/bpk-component-popover/stories.js:
```diff
@@ -36,44 +38,65 @@ const Paragraph = withDefaultProps(BpkText, {
+  ...$Diff<PopoverProps, { children: Node }>,
+  id: string,
+  changeProps: boolean,
+  targetFunction: ?Function,
```
Another place to refine this type

matthewdavidson replied:
Ended up with `targetFunction: ?() => ?HTMLElement,`

k0nserv commented on packages/bpk-component-popover/stories.js:
```diff
@@ -36,44 +38,65 @@ const Paragraph = withDefaultProps(BpkText, {
+type State = {
+  isOpen: boolean,
+  showLongContent: boolean,
+  changedTarget: ?Function,
```
Same as above

matthewdavidson commented:
@k0nserv, @shaundon have a look at my commit addressing your comments! https://github.com/Skyscanner/backpack/pull/435/commits/bd2ebf77df05e56a193a23adc32bcedee7504837

k0nserv commented:
>Yeah ... the return type is void where as these dummy functions were returning null.

Aren't they returning objects now though? Maybe we should have `mixed` for the return type instead?

matthewdavidson commented:
> Aren't they returning objects now though? Maybe we should have mixed for the return type instead?

Nope - returning an object would be `() => ({})`. I think it's fine as is.

k0nserv commented:
Right... Still think `mixed` might be better since we don't care

matthewdavidson commented:
Myself and @k0nserv agreed that for consumers convenience we'll used `mixed` return type. See https://github.com/Skyscanner/backpack/pull/435/commits/2df4dd0d249a269a2aeed62818bb877df5f7056e

matthewdavidson commented:
@k0nserv / @shaundon i need approvalz
